File: src/product/Stock.java
package product;

import java.util.HashMap;
import java.util.Map.Entry;
import utility.ToString;

public class Stock extends Product {
    private static Stock instance = null;
    private HashMap<Integer, Item> stockItem = new HashMap<>();
    private int nextAvailableID = 0;

    private Stock() {
        addItem(new Item("MLIK", 17));
        addItem(new Item("EGG", 12));
        addItem(new Item("TEA", 10));
        addItem(new Item("RICE", 2));
        addItem(new Item("PORK", 3));
        addItem(new Item("BROWNIE", 7));
        addItem(new Item("WATER", 1));
        addItem(new Item("MASK", 1));
        addItem(new Item("TISSUE", 7));

    }

    public static Stock getInstance() {
        if (instance == null) {
            instance = new Stock();
        }
        return instance;
    }

    @Override
    public void addItem(Item item) {
        int id = getItemID(item);
        if (id == nextAvailableID) {
            this.stockItem.put(id, item);
            nextAvailableID++;
        } else {
            int currentAmount = getAmount(id);
            this.stockItem.put(id, new Item(item.getName(), currentAmount + item.getAmount()));
        }
    }

    @Override
    public String getItems() {
        return ToString.stockItemToString(stockItem);
    }

    public boolean removeItem(Item item) {
        if (item.getAmount() <= checkStock(item)) {
            int id = getItemID(item);
            int currentItemInStock = getAmount(id);
            this.stockItem.put(id, new Item(item.getName(), currentItemInStock - item.getAmount()));
            return true;
        }
        return false;
    }

    public void removeProduct(int id) {
        this.stockItem.remove(id);
    }

    private int checkStock(Item item) {
        int id = getItemID(item);
        if (id == nextAvailableID) {
            return 0;
        }
        return stockItem.get(id).getAmount();
    }

    private int getItemID(Item item) {
        for (Entry<Integer, Item> entry : stockItem.entrySet()) {
            if (entry.getValue().equals(item)) {
                return entry.getKey();
            }
        }
        return nextAvailableID;
    }

    private int getAmount(int id) {
        return stockItem.get(id).getAmount();
    }

    public static void main(String[] args) {
        Stock stock = Stock.getInstance();
        System.out.println(stock.getItems());
        stock.addItem(new Item("TISSUE", 5));
        System.out.println(stock.getItems());
        stock.removeItem(new Item("WATER", 1));
        stock.removeItem(new Item("EGG", 3));
        stock.removeItem(new Item("RICE", 3));
        System.out.println(stock.getItems());
        stock.removeProduct(7);
        System.out.println(stock.getItems());
    }
}

File: TYPESCRIPT/src/product/Stock.ts
import { IProduct } from "./IProduct";
import { Item } from "./Item";

export class Stock implements IProduct {
  private static _instance: Stock = null;
  private _stockItem: Map<number, Item> = new Map();
  private _nextAvailableID: number = 0;

  private constructor() {
    this.addItem(new Item("MILK", 17));
    this.addItem(new Item("EGG", 12));
    this.addItem(new Item("TEA", 10));
    this.addItem(new Item("RICE", 2));
    this.addItem(new Item("PORK", 3));
    this.addItem(new Item("BROWNIE", 7));
    this.addItem(new Item("WATER", 1));
    this.addItem(new Item("MASK", 1));
    this.addItem(new Item("TISSUE", 7));
  }

  public static getInstance(): Stock {
    if (this._instance === null) {
      this._instance = new Stock();
    }
    return this._instance;
  }

  public addItem(item: Item): void {
    let id: number = this._getItemID(item);
    if (id === this._nextAvailableID) {
      this._stockItem.set(id, item);
      this._nextAvailableID++;
    } else {
      let currentAmount: number = this._getAmount(id);
      this._stockItem.set(
        id,
        new Item(item.getName, currentAmount + item.getAmount)
      );
    }
  }

  /**
   * TODO  change this after ToString util finished.
   */
  public getItems(): any {
    this._stockItem.forEach((element) => {
      console.log(element);
    });
    return this._stockItem;
  }

  public removeItem(item: Item): boolean {
    if (item.getAmount <= this._checkStock(item)) {
      let id: number = this._getItemID(item);
      let currentItemInStock: number = this._getAmount(id);
      this._stockItem.set(
        id,
        new Item(item.getName, currentItemInStock - item.getAmount)
      );
      return true;
    }
    return false;
  }

  public removeProduct(id: number): void {
    this._stockItem.delete(id);
  }

  private _checkStock(item: Item): number {
    let id: number = this._getItemID(item);
    if (id === this._nextAvailableID) {
      return 0;
    }
    return this._getAmount(id);
  }

  private _getItemID(item: Item): number {
    for (let entry of this._stockItem) {
      if (item.equals(entry[1])) {
        return entry[0];
      }
    }
    return this._nextAvailableID;
  }

  private _getAmount(id: number): number {
    return this._stockItem.get(id).getAmount;
  }
}
